alternator: Support `tablets_mode_for_new_keyspaces` config flag

ScyllaPiotr opened the pull request:
Until now, tablets in Alternator were experimental feature enabled only
when a TAG `experimental:initial_tablets` was present when creating a
table and associated with a numeric value.

After this patch, Alternator honours the value of
`tablets_mode_for_new_keyspaces` config flag.
Each table can be overridden to use tablets or not by supplying a new TAG
`system:initial_tablets`. The rules stay the same as with the earlier,
`experimental:initial_tablets` tag: when supplied with a numeric value, the table
will use tablets (as long as they are supported). When supplied with something
else (like a string "none"), the table will use vnodes. However, if
`tablets_mode_for_new_keyspaces` is set to `enforced` and a user requests
vnodes with the tag, an exception is thrown to the user.

Fixes #22463

Backport to 2025.4, it's important not to delay phasing out vnodes.

ScyllaPiotr reviewed (commented):
Thanks Radek for the review!

radoslawcybulski commented on alternator/executor.cc:
```diff
@@ -5883,11 +5887,19 @@ static lw_shared_ptr<keyspace_metadata> create_keyspace_metadata(std::string_vie
+                per_table_tablets = true;
+                initial_tablets = std::stol(it->second);
             } catch(...) {
+                per_table_tablets = false;
```
I do agree with not-fixing, someone wrote that spec for a reason, but could we get a message about this?

ScyllaPiotr replied:
> feels too easy to make an annoying mistake (imagine some newbie comes, gets the "simple" task of setting up a db and makes a typo, noone will even notice)

and

> could we get a message about this?

What message and where would you suggest?

The default is to not provide any extra tag and create the table with whatever the DB admin decided to use (by setting up the config flag `tablets_mode_for_new_keyspaces`). In order for the newbie to be able to make a mistake, they would need to learn about this special tag first. They would learn about it from the documentation and the usage is already clearly explained in the documentation. It is unlikely any newbie would read only the source code in order to learn how to create a table and miss the docs.
Don't get me wrong, I'm all in for the easiness of use, anticipating expectations and preventing misconceptions, it's just that the scenario you described does not seem very realistic to me.

ScyllaPiotr replied:
See https://github.com/scylladb/scylladb/pull/26211/files#diff-57dfb06b28a2c907df07874009eb9de541e5338704a9d733975f12be61b30c5fR5888 and other neighbouring info-level logging in the current version. Is this in-line with what you expected?

radoslawcybulski replied:
Excellent :) Just please add in liine 5897 at end end ", because of tag value is not an integer" (or something similar, your english is much better than mine).

ScyllaPiotr replied:
Done

radoslawcybulski commented on alternator/executor.cc:
```diff
@@ -5860,21 +5866,19 @@ static lw_shared_ptr<keyspace_metadata> create_keyspace_metadata(std::string_vie
+    // The value `enforced` for `default_tablets`, despite its name, can still
+    // be overridden (for Alternator) using the per-table tag.
+    static constexpr auto INITIAL_TABLETS_TAG_KEY = "system:initial_tablets";
+    std::optional<bool> per_table_tablets;
```
With value i was triple checking for what i was missing. :)

ScyllaPiotr replied:
Got ya. Fixed now.

radoslawcybulski commented on test/alternator/test_tablets.py:
```diff
@@ -76,29 +83,58 @@ def test_initial_tablets_number(dynamodb):
+# "tablets_mode_for_new_keyspaces": "enabled" and "enforced". The value
+# disabled is likely never to be used in the wild and additionally can
+# impact other tests running in parallel.
+def test_tablets_tag_vs_config(dynamodb):
```
Move to a test with private instance of scylla?
Setting to `enforced` might also blow up some other tests as well.

radoslawcybulski commented on alternator/executor.cc:
```diff
@@ -5860,21 +5866,19 @@ static lw_shared_ptr<keyspace_metadata> create_keyspace_metadata(std::string_vie
-    // to 0 enable tablets by default, with automatic number of tablets.
+    // The value `enforced` for `default_tablets`, despite its name, can still
+    // be overridden (for Alternator) using the per-table tag.
+    static constexpr auto INITIAL_TABLETS_TAG_KEY = "system:initial_tablets";
```
The problem i have with this is that it's not localized. You can type the tag manually (or someone can create the same) and access tag just fine. Tags are public, this is why i'd prefer all tags to be in single place for easier inspection, otherwise you need to search for `system:` at minimum and let's hope we won't get other prefixes.

ScyllaPiotr replied:
@nyh That is a valuable suggestion for future refactoring of all the string constants for TAGs.

ScyllaPiotr commented on docs/alternator/new-apis.md:
```diff
@@ -181,17 +181,19 @@ entire data center, or other data centers, in that case.
-Alternator tables default to using vnodes - not tablets.
+In this version, tablet support is almost complete, so new
+Alternator tables default to following what the global configuration flag
+[`tablets_mode_for_new_keyspaces`](../reference/configuration-parameters.html#confval-tablets_mode_for_new_keyspaces)
```
OK, it seems I figured this out myself. The links in the source need to refer to the real file names in the source.

ScyllaPiotr commented:
> What about failing tests?

Yes, fixing the tests and making them fixed are sometimes two different things...

nyh commented on alternator/executor.cc:
```diff
@@ -5850,7 +5860,8 @@ future<executor::request_return_type> executor::describe_continuous_backups(clie
 // manually create the keyspace to override this predefined behavior.
-static lw_shared_ptr<keyspace_metadata> create_keyspace_metadata(std::string_view keyspace_name, service::storage_proxy& sp, gms::gossiper& gossiper, api::timestamp_type ts, const std::map<sstring, sstring>& tags_map, const gms::feature_service& feat) {
+static lw_shared_ptr<keyspace_metadata> create_keyspace_metadata(std::string_view keyspace_name, service::storage_proxy& sp, gms::gossiper& gossiper, api::timestamp_type ts,
+            const std::map<sstring, sstring>& tags_map, const gms::feature_service& feat, const enum_option<db::tablets_mode_t>&& tablets_mode) {
```
ditto, I think tablets_mode could be a simple enum passed by value, not enum_option and not const (!?) rvalue reference.

nyh commented on alternator/executor.cc:
```diff
@@ -5860,32 +5871,43 @@ static lw_shared_ptr<keyspace_metadata> create_keyspace_metadata(std::string_vie
+    // Whether to use tablets for the table (actually for the keyspace
+    // of the table) is determined by the parameter `tablets_mode` (which is
+    // determined by `tablets_mode_for_new_keyspaces` config flag), as well as
+    // the presence and the value of a per-table tag below.
```
Instead of the vague "below" I would write "WRITE_ISOLATION_TAG_KEY" (or `system:initial_tablets`).

nyh commented on alternator/executor.cc:
```diff
@@ -5860,32 +5871,43 @@ static lw_shared_ptr<keyspace_metadata> create_keyspace_metadata(std::string_vie
-                initial_tablets = std::stol(tags_map.at(INITIAL_TABLETS_TAG_KEY));
+                // If the tag is a valid number, use it.
+                initial_tablets = std::stol(it->second);
+                elogger.info("Following {} tag, Alternator will attempt to create a keyspace {} with tablets, initially {} of them (0 means default number).", INITIAL_TABLETS_TAG_KEY, keyspace_name, initial_tablets.value());
```
I think "info" level is too much, it should be "trace". The thing is, you're not reporting here a rare occurance - you're actually tracing every single use of this tag.

nyh commented on alternator/executor.cc:
```diff
@@ -5860,32 +5871,43 @@ static lw_shared_ptr<keyspace_metadata> create_keyspace_metadata(std::string_vie
+                    throw api_error::validation(fmt::format("Creating keyspace {} with vnodes forbidden - tablets are enforced in configuration.", keyspace_name));
+                }
                 initial_tablets = std::nullopt;
+                elogger.info("Following {} tag containing non-numerical value, Alternator will attempt to create a keyspace {} with vnodes.", INITIAL_TABLETS_TAG_KEY, keyspace_name);
```
ditto, I think this should be trace level (personally, I would even drop it completely).

nyh commented on alternator/executor.cc:
```diff
@@ -5860,32 +5871,43 @@ static lw_shared_ptr<keyspace_metadata> create_keyspace_metadata(std::string_vie
+                // disable tablets and use vnodes by setting initial_tablets to a disengaged optional.
+                // When vnodes are asked for but tablets are enforced by config, throw.
+                if(tablets_mode == db::tablets_mode_t::mode::enforced) {
+                    throw api_error::validation(fmt::format("Creating keyspace {} with vnodes forbidden - tablets are enforced in configuration.", keyspace_name));
```
Please mention the name of the configuration variable that enforces this. This can be useful for users who will get this error.
I would also mention that "... with vnodes" was caused by setting the tag system:initial_tablets.

nyh commented on alternator/executor.cc:
```diff
@@ -5860,32 +5871,43 @@ static lw_shared_ptr<keyspace_metadata> create_keyspace_metadata(std::string_vie
+            // No per-table tag present, use the value from config
+            if(tablets_mode == db::tablets_mode_t::mode::enabled || tablets_mode == db::tablets_mode_t::mode::enforced) {
+                initial_tablets = 0;
+                elogger.info("Following the `tablets_mode_for_new_keyspaces` flag from the settings, Alternator will attempt to create a keyspace {} with tablets.", keyspace_name);
```
again I would use trace level for all these log messages, or drop them entirely.

nyh commented on alternator/executor.cc:
```diff
@@ -5860,32 +5871,43 @@ static lw_shared_ptr<keyspace_metadata> create_keyspace_metadata(std::string_vie
+            }
+        } else {
+            // No per-table tag present, use the value from config
+            if(tablets_mode == db::tablets_mode_t::mode::enabled || tablets_mode == db::tablets_mode_t::mode::enforced) {
```
nitpick: "!= mode::disabled" is shorter and faster :-)
(assuming the "-1" setting doesn't really exist?)

nyh commented on alternator/executor.cc:
```diff
@@ -5860,32 +5871,43 @@ static lw_shared_ptr<keyspace_metadata> create_keyspace_metadata(std::string_vie
+                initial_tablets = 0;
+                elogger.info("Following the `tablets_mode_for_new_keyspaces` flag from the settings, Alternator will attempt to create a keyspace {} with tablets.", keyspace_name);
+            } else {
+                elogger.info("Following the `tablets_mode_for_new_keyspaces` flag from the settings, Alternator will attempt to create a keyspace {} with vnodes.", keyspace_name);
```
I see you didn't put `initial_tablets = std::nullopt;`. I agree that it's not needed, because that's the default value of this variable, but then we also don't need to do it above, right?

nyh commented on docs/alternator/new-apis.md:
```diff
@@ -181,27 +181,30 @@ entire data center, or other data centers, in that case.
-tablets are smaller pieces of tables that are easier to move between nodes,
-and allow for faster growing or shrinking of the cluster when needed.
+It replaces the older approach which was named "vnodes". See
+[Data Distribution with Tablets](../architecture/tablets.rst) for details.
```
Ok, but why did you decide to remove my two-line explanation of what tablets are?
I agree that a few years from now, when tablets are the default (or even the only way), the Alternator documentation doesn't need to explain tablets at all. But we're not there yet, and this is why this section even exists in this document.

nyh replied:
I have a pet peeve about  people claiming that because documentation can get out of sync, it shouldn't be written at all.
But I think it's good enough for now.

nyh commented on docs/alternator/new-apis.md:
```diff
@@ -181,27 +181,30 @@ entire data center, or other data centers, in that case.
+  an exception will be thrown.
 
-The `experimental:initial_tablets` tag only has any effect while creating
+The `system:initial_tablets` tag has only effect while creating
```
This sentence was better before your modification...

I'll let Copilot explain why:
In English, "only" should be placed as close as possible to the word or phrase it is modifying. In the first sentence, "only" modifies "has any effect," which is the intended meaning. "Has any effect" is a standard way to say "to have an impact or influence." "Has only effect" is not idiomatic and would sound awkward to native speakers.

nyh commented on test/alternator/test_tablets.py:
```diff
@@ -84,6 +91,37 @@ def test_initial_tablets_number(dynamodb):
+# The test checks only 2 of the values of the config flag
+# "tablets_mode_for_new_keyspaces": "enabled" and "enforced". The value
+# "disabled" is likely never to be used in the wild and additionally
+# can impact other tests running in parallel.
```
There are no "other tests running in parallel" - we already assume this in other places and basically any use of scylla_config_temporary assumes that.
Note that "in parallel" means on the same Scylla instance. It's perfectly fine for test.py to run multiple tests in parallel because it runs them against difference instances of Scylla.

nyh commented on test/alternator/test_tablets.py:
```diff
@@ -84,6 +91,37 @@ def test_initial_tablets_number(dynamodb):
+# then if the tag asks for vnodes, an exception is thrown.
+# The test checks only 2 of the values of the config flag
+# "tablets_mode_for_new_keyspaces": "enabled" and "enforced". The value
+# "disabled" is likely never to be used in the wild and additionally
```
I think instead of explaining why you don't want to test "disabled" mode, you should just test it - it's easy.

ScyllaPiotr reviewed (commented):
Thanks for the review @nyh . The requested changes have been introduced, let's wait for the tests to finish.
You are free to do the re-review, as soon as you find the time for it.

ScyllaPiotr commented:
> Failed Tests (26/49591):
> - [test_gsi_describe_indexstatus.debug.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/20671/testReport/junit/alternator/test_gsi/debug___test_py___debug___test_gsi_describe_indexstatus_debug_1) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_gsi_describe_indexstatus.debug.1)
> - [test_gsi_describe_indexstatus.debug.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/20671/testReport/junit/alternator/test_gsi/debug___test_py___debug___test_gsi_describe_indexstatus_debug_2) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_gsi_describe_indexstatus.debug.2)
> - [test_gsi_describe_indexstatus.debug.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/20671/testReport/junit/alternator/test_gsi/debug___test_py___debug___test_gsi_describe_indexstatus_debug_3) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_gsi_describe_indexstatus.debug.3)
> - [test_gsi_describe_indexstatus.dev.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/20671/testReport/junit/alternator/test_gsi/dev___test_py___dev___test_gsi_describe_indexstatus_dev_1) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_gsi_describe_indexstatus.dev.1)
> - [test_gsi_describe_indexstatus.dev.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/20671/testReport/junit/alternator/test_gsi/dev___test_py___dev___test_gsi_describe_indexstatus_dev_2) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_gsi_describe_indexstatus.dev.2)
> - [test_gsi_describe_indexstatus.dev.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/20671/testReport/junit/alternator/test_gsi/dev___test_py___dev___test_gsi_describe_indexstatus_dev_3) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_gsi_describe_indexstatus.dev.3)
> - [test_gsi_describe_indexstatus.release.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/20671/testReport/junit/alternator/test_gsi/release___test_py___release___test_gsi_describe_indexstatus_release_1) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_gsi_describe_indexstatus.release.1)
> - [test_gsi_describe_indexstatus.release.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/20671/testReport/junit/alternator/test_gsi/release___test_py___release___test_gsi_describe_indexstatus_release_2) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_gsi_describe_indexstatus.release.2)
> - [test_gsi_describe_indexstatus.release.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/20671/testReport/junit/alternator/test_gsi/release___test_py___release___test_gsi_describe_indexstatus_release_3) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_gsi_describe_indexstatus.release.3)
> - [test_gsi_delete_with_lsi.debug.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/20671/testReport/junit/alternator/test_gsi_updatetable/debug___test_py___debug___test_gsi_delete_with_lsi_debug_1) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_gsi_delete_with_lsi.debug.1)

All the tests fail on `wait_for_gsi` which times out after waiting for 10 mins. `wait_for_gsi` continues looping because `index_status` is not moving on from `CREATING`.
The problem reproduces on my machine.
As soon as I order a table with vnodes (instead of the now-default tablets), the tests pass.
The tests pass also on a version (a commit on master) that is a root of my branch.
This is where I got until now.

@nyh Might it be that executor incorrectly sets values of `alternator::table_status` when used with tablets?

swasik commented:
@ScyllaPiotr is thin `index_status`  something Alternator specific? If I understand correctly tests pass on the commit from which you branched so there has to be some regression recently. Am I correct? @piodul do you know if there were some changes to indexes recently that can affect this?

swasik commented:
> > The tests pass also on a version (a commit on master) that is a root of my branch.
> 
> @ScyllaPiotr When running the test on current master, does it use vnodes or tablets?

@piodul it passes with vnodes and fails with tablets.

nyh commented:
> > Please rebase and try again.
> 
> No, this did not help. Will investigate changing of the value of the `index_status` flag.

I implemented IndexStatus in commit cea7aacc52120e76b017c9a29773d9a439bca059: This is the DynamoDB API for the user to know when the GSI becomes available, and our implementation need to figure out whether the materialized view (which implements the GSI) has completed its building.

The implementation is kind of nasty (see the aforementioned commit, its fairly heavily commented), and highly depends on the implementation of the "view building code" and how it tells us when if a view has finished building. Recently, @piodul's team was very active on completely rewriting the view building code, so I suspect if you had problems with it just for tablets mode, then the implementation of the view building code - or specifically - the metadata used to tell when a specific view is already fully built - has changed for tablets, and you'll need to find out what is the new way to check if a specific materialized view has already finished building.

nyh commented:
By the way, in the comment I added in https://github.com/scylladb/scylladb/commit/cea7aacc52120e76b017c9a29773d9a439bca059 I mentioned that:

> Currently we don't have in view_ptr the knowledge whether a view finished building long ago - so checking this involves a somewhat inefficient, but still node-local, process:

I really hope that when the view building code was rewritten, this "wish" came true. The fact that view building has completed for a view in the schema is permanent - once it completed building it never needs to build again, even if new nodes are added to the cluster or anything like that. So the "view building completed" should have been a permanent flag on the view's schema, and as any permanent flag in the schema it would have a copy in memory that we could inspect immediately. This is what we **should** have had. But when I wrote that IndexStatus code we didn't - which is how we ended up with the ugly code in that commit :-(

nyh commented:
> To supplement what Nadav wrote, both failing tests:

I saw more than just these two tests failing. Indeed, in addition to the add-a-GSI tests that failed, we have additional tests that check that the empty view created together with the table also completes building. Remember that even when the base table is absolutely empty, we do NOT currently have a shortcut that avoids the view building - and the view building runs and scans the empty table (it might not even be empty, if the test immediately begins to write to it).

By the way, why do we NOT have such a shortcut for creating a view on an empty base table? The thing is that in CQL, there is no operation like Alternator's CreateTable which can create an empty base table together with a view. So in CQL, you can never really know if the base table is empty (the user may write to the base table in parallel with creating 

By the way, @Jadw1 - we discussed exactly this issue above - we don't start the view builder on CreateTable. In fact, we don't need to start the view builder at all on CreateTable - because we know the new table must be empty! BUT, in that case we do need to write down that the view building has completed.

Jadw1 commented:
@nyh @piodul
If we are sure that the base table is always empty, maybe we can just mark the view as built in `system.view_build_status_v2` and skip view building entirely?

In old view builder, information in `system.built_views` and `system.view_building_status_v2` tables weren't "connected".

But with tablets and the coordinator, when a view is built, it is marked as SUCCESS on all nodes at once in `system.view_build_status_v2`.
 Then, basing on this information, each node will create an entry in `system.built_views` (`view_building_worker::update_built_views()`).

Jadw1 commented:
FYI: Posted https://github.com/scylladb/scylladb/pull/26657. Also, I've rebased this PR on top of mine and checked failing `test_gsi` and `test_gsi_updatetable` tests locally. They are working.

ScyllaPiotr commented:
A moderately painful rebase later, I have a version that is indeed working locally! So let's wait for the CI to agree with my local build.

ScyllaPiotr commented:
I will not pick this up this moment, so in the meantime let's allow the CI to try again to see which of the problems are persistent.
@scylladbbot trigger-ci

ScyllaPiotr commented:
> Failed Tests (10/77503):
>
> * [test_ttl_expiration_streams[using tablets].debug.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/21460/testReport/junit/alternator/test_ttl/debug___test_py___debug___test_ttl_expiration_streams_using_tablets__debug_1)
> * ...

New day, new failing test...
Rebased and fixed.

BTW the failing code did not live on my branch and it took me a moment to realize, that CI rebases my PR onto **current** master before building and running the tests.

nyh commented on alternator/executor.cc:
```diff
@@ -5929,22 +5941,20 @@ future<executor::request_return_type> executor::describe_continuous_backups(clie
-    // used by default on new Alternator tables. Change this initialization
-    // to 0 enable tablets by default, with automatic number of tablets.
+    // When vnodes are asked for by the tag value, but tablets are enforced by config,
+    // throw an exception to the client.
```
I think this check is actually missing...

Maybe you meant is that something inside the keyspace-creation call chain checks the tablets mode on its own throws an exception if you try to create a table with vnodes. But this unknown exception will be reported like an ugly "internal server error" - and we want it to be a clear InvalidRequest with a clear message.

I see you actually wrote a test for this, test_tablets.py::test_tablets_tag_vs_config, but it doesn't work because tablets_mode_for_new_keyspaces is not live-updateable :-( I don't making that test "xfail" was a good idea. I think you should have either made tablets_mode_for_new_keyspaces live-updatable, or if there was a reason why NOT to do this, the test should be moved to test/cluster, where a new Scylla node can be started with specific options. I'll do one of these in my followup patch.

nyh commented on docs/alternator/new-apis.md:
```diff
@@ -181,27 +181,30 @@ entire data center, or other data centers, in that case.
-  without tablets - i.e., using vnodes.
+  without tablets - i.e., using vnodes. However, when vnodes are asked for by the tag value,
+  but tablets are `enforced` by the `tablets_mode_for_new_keyspaces` configuration flag,
+  an exception will be thrown.
```
nitpick: "exception will be thrown" is not the appropriate terminology for the DynamoDB API. We're not talking about C++ code here.

nyh commented on test/alternator/test_scan.py:
```diff
@@ -414,29 +414,38 @@ def test_scan_long_partition_tombstone_string(dynamodb, query_tombstone_page_lim
 # For single-partition scans, the page size is more exactly 1 MB.
-def test_scan_paging_missing_limit(dynamodb):
+#
+# The test is run twice, with two different values of N (the number of items in a batch write).
```
The new comment and the commit message mix up "page size", "batch size" and even uses the phrase "batch write" for the table size (note that the AWS SDK does **not** write the entire thing in a single batch, it's just a loop which collects several smaller batches). I'll fix it up myself.

nyh commented on test/alternator/test_scan.py:
```diff
@@ -414,29 +414,38 @@ def test_scan_long_partition_tombstone_string(dynamodb, query_tombstone_page_lim
+@pytest.mark.parametrize("N", [
+    pytest.param(200, marks=pytest.mark.xfail(
+        reason="Issue #10327: small tables don't respect paging limits")),
+    1400
```
I don't think we need to run the the test with the smaller number. The higher number already proves the "missing limit" works correctly (doesn't scan the entire table).
I am not happy we need to increase 600 to 1400 - it can make the test almost 3 times slower. I'll check myself if a lower number works reasonably and if it does, I'll lower it. I don't see any reason why we need to support 16 tablets per shard and not a different number.

nyh replied:
I think 14 MB is a bit excessive, and will make this test noticably slower than it was. I checked experimentally that up to initial_tablets=8 (so 4 tablets per shard, like you said was the default in debug mode), 8 MB was also enough so I want to use that.

annastuchlik commented on docs/alternator/new-apis.md:
```diff
@@ -181,27 +181,30 @@ entire data center, or other data centers, in that case.
-Alternator tables default to using vnodes - not tablets.
+In this version, tablet support is almost complete, so new
+Alternator tables default to following what the global configuration flag
+[`tablets_mode_for_new_keyspaces`](../reference/configuration-parameters.rst#confval-tablets_mode_for_new_keyspaces)
```
You got me here. Indeed, the link is working, but I didn't realize that until I hovered exactly over it.
Similarly, a user might find it hard to guess that it's a link, as underlining and coloring are not supported when backquotes are used in links in our  MD files. For the sake of user experience, I think it would be better to make it look like a link at first glance. 
Anyway, I'll approve this PR as there are no obvious issues to fix. Thanks.
